### PR TITLE
fix: allow cache to store proper token in case cache ininvalid JSON

### PR DESCRIPTION
**Pull request by gabrielfs7**
Task https://oat-sa.atlassian.net/browse/ADF-959

In case an invalid token JSON was saved in the cache, we get correct one from DB and store it in the cache again.

### CODE REVIEW

File: model/storage/ConsumerStorage.php [hunk 1]
```diff
@@ -73,21 +77,52 @@ public function setConsumerToken(\core_kernel_classes_Resource $consumer, Access
      *
      * Fetch token from cache if exists, otherwise set it
      *
-     * @param $tokenHash
+     * @param string $tokenHash
```
Comment (gabrielfs7) [inline model/storage/ConsumerStorage.php:80]: I wont add typehint and return type in the scope of this PR, cause we have mixed behavior relying in boolean and other operators in different places.

File: model/storage/ConsumerStorage.php [hunk 2]
```diff
@@ -14,12 +15,15 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  *
- * Copyright (c) 2018 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
- *
+ * Copyright (c) 2018-2021 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```
Comment (shpran) [inline model/storage/ConsumerStorage.php:18]:
```suggestion
 * Copyright (c) 2018-2022 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

File: model/storage/ConsumerStorage.php [hunk 3]
```diff
@@ -73,21 +77,52 @@ public function setConsumerToken(\core_kernel_classes_Resource $consumer, Access
...
      * Fetch token from cache if exists, otherwise set it
      *
-     * @param $tokenHash
+     * @param string $tokenHash
      * @return AccessToken
-     * @throws \common_Exception
-     * @throws \common_exception_NotFound
-     * @throws \core_kernel_persistence_Exception
+     * @throws common_Exception
+     * @throws common_exception_NotFound
```
Comment (shpran) [inline model/storage/ConsumerStorage.php:83]: No need to add throw `NotFound` phpdoc, as we throw it inside this method

File: model/storage/ConsumerStorage.php [hunk 4]
```diff
@@ -112,16 +147,16 @@ public function getConsumer($clientKey, $clientSecret)
...
     }
 
     /**
      * Get an oauth consumer by token hash
      *
      * @param $hash
-     * @return \core_kernel_classes_Resource
-     * @throws \common_exception_NotFound
+     * @return core_kernel_classes_Resource
+     * @throws common_exception_NotFound
```
Comment (shpran) [inline model/storage/ConsumerStorage.php:159]: Unnecessary phpdoc (`throws`)

**Review by shpran — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by bartlomiejmarszal — COMMENTED** [review]
This is great refactoring that also includes wrong decoded json exception that will surely allow us to recognise issue that is happening on prod. It's not addressing this issue thought.

**Review by bartlomiejmarszal — COMMENTED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: model/storage/ConsumerStorage.php (hunk 3 above)
Comment (gabrielfs7) [reply model/storage/ConsumerStorage.php:83]: The PHP doc is for the clients of this class, not for the developer of the class. So in theory we should add all the possible exception that this class throws.

**Review by bartlomiejmarszal — APPROVED** [review]

**Review by hectoras — APPROVED** [review]

**Comment by gabrielfs7** [comment]
> This is great refactoring that also includes wrong decoded json exception that will surely allow us to recognise issue that is happening on prod. It's not addressing this issue thought.

It is actually making sure that if a wrong token was persisted, it will get the proper one from DB and store it properly on the cache.

It is also fixing a previous object serialization save done before and saving the proper json encoded as it should be https://github.com/oat-sa/extension-tao-oauth/pull/62/files#diff-53e27a8f69e859a6bb9cd249281f41a5b2ca4c590c97edfb18bf217f87fb34f3L89

File: model/storage/ConsumerStorage.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (gabrielfs7) [reply model/storage/ConsumerStorage.php:18]: back to the future :) I will fix all of them. Thanks

Comment (gabrielfs7) [reply model/storage/ConsumerStorage.php:18]: https://github.com/oat-sa/extension-tao-oauth/pull/62/files#r810936821

**Comment by AntonNeviarouski** [comment]
QA verified

File: model/storage/ConsumerStorage.php (hunk 3 above)
Comment (gabrielfs7) [reply model/storage/ConsumerStorage.php:83]: We decided to make it optional, so I am keeping them :)

**Pull request merged**